storage/replica.go

```diff
@@ -1421,20 +1421,25 @@ func (r *Replica) handleRaftReady() error {
 	batch := r.store.Engine().NewBatch()
 	defer batch.Close()
 	if !raft.IsEmptySnap(rd.Snapshot) {
+		// TODO(peter): Could use a "distinct" batch within applySnapshot, but we
+		// also need the Batch.Defer mechanism.
 		var err error
 		if lastIndex, err = r.applySnapshot(batch, rd.Snapshot); err != nil {
 			return err
 		}
 		// TODO(bdarnell): update coalesced heartbeat mapping with snapshot info.
 	}
+	// We know that all of the writes from here forward will be to distinct keys.
```

Comment:
Well, this was mostly straightforward except for one large gotcha: the current usage of `Distinct()` within `Replica.{Put,ConditionalPut}` would now cause a flush on every request which completely destroys the distinct optimization.